keps/sig-node/375-cpu-manager/README.md

```diff
@@ -323,6 +395,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+###### Does enabling the feature change any default behavior?
+
+No, unless the non-none policy is explicitly configured.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, using the kubelet config.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The impact is node-local only.
+If the state of a node is steady, no changes.
+If a guaranteed pod is admitted, running non-guaranteed pods will have their CPU cgroup changed while running.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes, covered by e2e tests
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+A rollout can fail if a bug in the cpumanager prevents _new_ pods to start, or existing pods to be restarted.
+Already running workload will not be affected if the node state is steady
+
+###### What specific metrics should inform a rollback?
+
+Pod creation errors o a node-by-node basis.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No to both.
+Changes in behavior only affects pods meeting the conditions (guaranteed QoS, integral CPU request) scheduler after the upgrade.
+Running pods will be unaffected by any change. This offers some degree of safety in both upgrade->rollback
+and upgrade->downgrade->upgrade scenarios.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+Monitor the pod admission counter
+Monitor the pods not going running after successful schedule
+
+###### How can an operator determine if the feature is in use by workloads?
+
+The operator need to inspect the node and verify the cpu pinning assignment either checking the cgroups on the node
+or accessing the podresources API of the kubelet.
+
+###### How can someone using this feature know that it is working for their instance?
+
+
+- [X] Other (treat as last resort)
+  - Details: the containers need to check the cpu set they are allowed to run; in addition, node agents (e.g. node_exporter)
+    can report the CPU assignment
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+- N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [ ] Other (treat as last resort)
+  - Details:
+     a operator should check that pods go running correctly and the cpu pinning is performed. The latter can
+     be checked by inspecting the cgroups at node level.
```

Comment:
Where is this exposed in aggregate?  if it's exposed by something that can be scraped by metrics, that's ok, but we do look for at least one existent implementation of such exposure.

Reply:
to the best of my knowledge, no well known software component do this, but I'll doublecheck (e.g. node_exporter) and add here as comment for historical record.

Reply:
> to the best of my knowledge, no well known software component do this, but I'll doublecheck (e.g. node_exporter) and add here as comment for historical record.

Being able to know whether a per-node feature is working at scale required for PRR sign off.  Even a simple mis-pinned metric would be sufficient, but ssh-ing to every node is impractical.

Separately, how does a pod author know that his pod is pinned correctly?

Reply:
Thanks for raising. Point taken, checking. Lack of status reporting can be a gap. Checking and reporting.

Reply:
cadvisor should have the cpu set information for the pod.

Reply:
I think we can add two new metrics in `k/k/pkg/kubelet/metrics` to track
1. the total allocation performed by cpu manager which will cause cpu pinning (`cpu_pinning_requests_total`?)
2. the portion of the above which failed (`cpu_pinning_requests_errors_total`?)

this way the cluster admin can have a feedback to check if the cpu pinning feature works and how many allocation failures there are (usually zero), without exposing any unnecessary node-level machine config (e.g. which pinned cores)

could look like this:
```
diff --git a/pkg/kubelet/metrics/metrics.go b/pkg/kubelet/metrics/metrics.go
index 17267b888d7..c8c3a03eaeb 100644
--- a/pkg/kubelet/metrics/metrics.go
+++ b/pkg/kubelet/metrics/metrics.go
@@ -86,6 +86,10 @@ const (
 	// Metrics to track ephemeral container usage by this kubelet
 	ManagedEphemeralContainersKey = "managed_ephemeral_containers"
 
+	// Metrics to track the CPU manager behavior
+	CPUManagerPinningRequestsTotalKey = "cpu_manager_pinning_requests_total"
+	CPUManagerPinningErrorsTotalKey   = "cpu_manager_pinning_errors_total"
+
 	// Values used in metric labels
 	Container          = "container"
 	InitContainer      = "init_container"
@@ -497,6 +501,26 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		},
 	)
+
+	// CPUManagerPinningRequestsTotal tracks the number of times the pod spec will cause the cpu manager to pin cores
+	CPUManagerPinningRequestsTotal = metrics.NewCounter(
+		&metrics.CounterOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           CPUManagerPinningRequestsTotalKey,
+			Help:           "The number of cpu core allocations which required pinning.",
+			StabilityLevel: metrics.ALPHA,
+		},
+	)
+
+	// CPUManagerPinningErrorsTotal tracks the number of times the pod spec required the cpu manager to pin cores, but the allocation failed
+	CPUManagerPinningErrorssTotal = metrics.NewCounter(
+		&metrics.CounterOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           CPUManagerPinningErrorssTotalKey,
+			Help:           "The number of cpu core allocations which required pinning failed.",
+			StabilityLevel: metrics.ALPHA,
+		},
+	)
 )
 
 var registerMetrics sync.Once
```
(names and descriptions subjected to change). I'll file a separate demo PR and update this one ASAP, since I think this is a nice monitoring improvement anyway.

Reply:
implementation sketch: https://github.com/kubernetes/kubernetes/pull/112855

Reply:
Let's capture these proposed metrics in the KEP. If @derekwaynecarr and @deads2k are okay with it we can get https://github.com/kubernetes/kubernetes/pull/112855 merged before 1.26 code freeze. 
We can gather feedback from other reviewers regarding the metrics to be gathered (and naming) as part of the KEP review. I am happy with the approach so far.

Reply:
> Let's capture these proposed metrics in the KEP. If @derekwaynecarr and @deads2k are okay with it we can get [kubernetes/kubernetes#112855](https://github.com/kubernetes/kubernetes/pull/112855) merged before 1.26 code freeze. We can gather feedback from other reviewers regarding the metrics to be gathered (and naming) as part of the KEP review. I am happy with the approach so far.

sure thing. I'll update the KEP proper ASAP.

Reply:
The metrics described in this thread look good to me.  If you update the KEP with these names (even before the impl merges), that will do.

Reply:
thanks, on it.